library/core/src/str/mod.rs

```diff
@@ -445,15 +451,23 @@ impl str {
     /// ```
     #[unstable(feature = "round_char_boundary", issue = "93743")]
     #[inline]
-    pub fn ceil_char_boundary(&self, index: usize) -> usize {
+    pub const fn ceil_char_boundary(&self, index: usize) -> usize {
         if index >= self.len() {
             self.len()
         } else {
-            let upper_bound = Ord::min(index + 4, self.len());
-            self.as_bytes()[index..upper_bound]
-                .iter()
-                .position(|b| b.is_utf8_char_boundary())
-                .map_or(upper_bound, |pos| pos + index)
+            let mut i = index;
+            while i < self.len() {
+                if self.as_bytes()[i].is_utf8_char_boundary() {
+                    break;
+                }
+                i += 1;
+            }
+
+            //  The character boundary will be within four bytes of the index
+            debug_assert!(i <= index + 3);
+            debug_assert!(self.is_char_boundary(i));
```

Comment:
This debug-assert (and its partner above) is pretty unnecessary because it's basically verifying that `self.as_bytes()[i].is_utf8_char_boundary() || i == self.len()` is equivalent to `self.is_char_boundary(i)`, which is practically the definition of that method.